src/cargo/ops/cargo_update.rs

```diff
@@ -207,6 +214,251 @@ pub fn print_lockfile_changes(
+            name
+        );
+        return Ok(dependency);
+    };
+
+    let upgrade_message = format!("{} {} -> {}", name, current, new_req_string);
+    trace!(upgrade_message);
+
+    if upgrade_messages.insert(upgrade_message.clone()) {
+        opts.gctx
+            .shell()
+            .status_with_color("Upgrading", &upgrade_message, &style::GOOD)?;
+    }
+
+    upgrades.insert(name.to_string(), latest.clone());
+
+    let req = OptVersionReq::Req(VersionReq::parse(&latest.to_string())?);
+    let mut dep = dependency.clone();
+    dep.set_version_req(req);
+    Ok(dep)
+}
+
+/// Update manifests with upgraded versions, and write to disk. Based on cargo-edit.
+/// Returns true if any file has changed.
+pub fn write_manifests(
+    ws: &Workspace<'_>,
+    opts: &UpdateOptions<'_>,
+    upgrades: &HashMap<String, Version>,
+) -> CargoResult<bool> {
+    if upgrades.is_empty() {
+        return Ok(false);
+    }
+
+    let mut any_file_has_changed = false;
+
+    let manifest_paths = std::iter::once(ws.root_manifest())
+        .chain(ws.members().map(|member| member.manifest_path()))
+        .collect::<Vec<_>>();
+
+    for manifest_path in manifest_paths {
+        trace!(
+            "updating TOML manifest at `{:?}` with upgraded dependencies",
+            manifest_path
+        );
+
+        let crate_root = manifest_path
+            .parent()
+            .expect("manifest path is absolute")
+            .to_owned();
+
+        let mut local_manifest = LocalManifest::try_new(&manifest_path)?;
+        let mut manifest_has_changed = false;
+
+        for dep_table in local_manifest.get_dependency_tables_mut() {
+            for (mut dep_key, dep_item) in dep_table.iter_mut() {
+                let dep_key_str = dep_key.get();
+                let dependency = match crate::util::toml_mut::dependency::Dependency::from_toml(
+                    &manifest_path,
+                    dep_key_str,
+                    dep_item,
+                ) {
+                    Ok(dependency) => dependency,
+                    Err(err) => {
+                        opts.gctx
+                            .shell()
+                            .warn(&format!("ignoring {dep_key}, unsupported entry: {err}"))?;
+                        continue;
+                    }
+                };
+
+                let Some(current) = dependency.version() else {
+                    trace!("skipping dependency without a version: {}", dependency.name);
+                    continue;
+                };
+
+                let Some(latest) = upgrades.get(&dependency.name) else {
+                    trace!(
+                        "skipping dependency without an upgrade: {}",
+                        dependency.name
+                    );
+                    continue;
+                };
+
+                let Some(new_req_string) = upgrade_requirement(current, latest)? else {
```

Comment:
Why do we call `upgrade_requirement in both `write_manifests` and `upgrade_manifests`?

Reply:
`write_manifests` needs to know what version req to put in the manifest file. `upgrade_manifests` just needs it for the upgrade message. If we could use the latest version rather than the new version req in the upgrade message, we wouldn't need to call it. Also, we cannot reuse the output from the first in the second, because `write_manifests` will call it with the actual syntax in the TOML file and preserve the formatting, while `upgrade_manifests` doesn't know about any formatting and will call it with a stringified version req.